tools/dotnet-linker/AppBundleRewriter.cs

```diff
@@ -0,0 +1,1084 @@
+						&& v.Parameters [0].ParameterType.Is ("System", "IntPtr")
+						&& !v.HasGenericParameters);
+			}
+		}
+
+		public MethodReference Runtime_RetainNSObject {
+			get {
+				return GetMethodReference (PlatformAssembly, ObjCRuntime_Runtime, "RetainNSObject", (v) =>
+						v.IsStatic
+						&& v.HasParameters
+						&& v.Parameters.Count == 1
+						&& v.Parameters [0].ParameterType.Is ("Foundation", "NSObject")
+						&& !v.HasGenericParameters);
+			}
+		}
+
+		public MethodReference Runtime_RetainNativeObject {
+			get {
+				return GetMethodReference (PlatformAssembly, ObjCRuntime_Runtime, "RetainNativeObject", (v) =>
+						v.IsStatic
+						&& v.HasParameters
+						&& v.Parameters.Count == 1
+						&& v.Parameters [0].ParameterType.Is ("ObjCRuntime", "INativeObject")
+						&& !v.HasGenericParameters);
+			}
+		}
+
+		public MethodReference Runtime_RetainAndAutoreleaseNSObject {
+			get {
+				return GetMethodReference (PlatformAssembly, ObjCRuntime_Runtime, "RetainAndAutoreleaseNSObject", (v) =>
+						v.IsStatic
+						&& v.HasParameters
+						&& v.Parameters.Count == 1
+						&& v.Parameters [0].ParameterType.Is ("Foundation", "NSObject")
+						&& !v.HasGenericParameters);
+			}
+		}
+
+		public MethodReference Runtime_RetainAndAutoreleaseNativeObject {
+			get {
+				return GetMethodReference (PlatformAssembly, ObjCRuntime_Runtime, "RetainAndAutoreleaseNativeObject", (v) =>
+						v.IsStatic
+						&& v.HasParameters
+						&& v.Parameters.Count == 1
+						&& v.Parameters [0].ParameterType.Is ("ObjCRuntime", "INativeObject")
+						&& !v.HasGenericParameters);
+			}
+		}
+
+		public MethodReference UnmanagedCallersOnlyAttribute_Constructor {
+			get {
+				return GetMethodReference (CorlibAssembly, "System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute", ".ctor", (v) => v.IsDefaultConstructor ());
+			}
+		}
+
+		public MethodReference Unsafe_AsRef {
+			get {
+				return GetMethodReference (CorlibAssembly, "System.Runtime.CompilerServices.Unsafe", "AsRef", (v) =>
+						v.IsStatic
+						&& v.HasParameters
+						&& v.Parameters.Count == 1
+						&& v.Parameters [0].ParameterType.IsPointer
+						&& v.Parameters [0].ParameterType.GetElementType ().Is ("System", "Void")
+						&& v.HasGenericParameters);
+			}
+		}
+
+		public void SetCurrentAssembly (AssemblyDefinition value)
+		{
+			current_assembly = value;
+		}
+
+		public void SaveCurrentAssembly ()
+		{
+			SaveAssembly (CurrentAssembly);
+		}
+
+		void SaveAssembly (AssemblyDefinition assembly)
+		{
+			if (assembly != CurrentAssembly && assembly != PlatformAssembly)
+				throw new InvalidOperationException ($"Can't save assembly {assembly.Name} because it's not the current assembly ({CurrentAssembly.Name}) or the platform assembly ({PlatformAssembly.Name}).");
+			var annotations = configuration.Context.Annotations;
+			var action = annotations.GetAction (assembly);
+			if (action == AssemblyAction.Copy)
+				annotations.SetAction (assembly, AssemblyAction.Save);
```

Comment:
```suggestion
			if (action == AssemblyAction.Copy) {
				// Preserve TypeForwardedTo which would the linker sweep otherwise
				if (assembly.MainModule.HasExportedTypes) {
					foreach (var type in assembly.MainModule.ExportedTypes) {
						annotations.Mark (type);
					}
				}
				annotations.SetAction (assembly, AssemblyAction.Save);
			}
```

This resolves the issue in https://github.com/xamarin/xamarin-macios/pull/18268#issuecomment-1545567672. It workarounds the quirk here - https://github.com/dotnet/runtime/blob/9dd59af3aee2f403e63887afef50d98022a2e575/src/tools/illink/src/linker/Linker.Steps/SweepStep.cs#L191-L200. It may be cleaner to resolve it on the linker side but that's problematic short term.

Reply:
Correct. In this case the original action is `copy`, MSR changes it to `save` and the side-effect is that the exported types (TypeForwardedTo) is dropped. Marking the exported types results in both the type-forwarders and MSR in the output assembly.